runtime/lua/vim/lsp/inlay_hint.lua

```diff
@@ -98,6 +98,91 @@ function M.on_refresh(err, _, ctx, _)
   return vim.NIL
 end
 
+--- @class InlayHintOptions
+--- @field bufnr number|nil
+--- @field filter function|nil
+--- @field id number|nil
+--- @field name string|nil
+--- @field range { start: number[], end: number[] }|nil
+
+--- Returns inlay hints in the (optionally) given buffer, client, and range.
+--- This method does not make an LSP request, but just queries locally stored inlay hints.
+---
+--- @param options table|nil Optional table which holds the following optional fields:
+--- - bufnr (number|nil):
+---     Restrict hints to the clients attached to the given buffer, defaults to the current
+---     buffer (0).
```

Comment:
I'd tend to change `nil` to imply all buffers. Otherwise you've no way to get all inlay hints. While getting the one for the current buffer could be done using `{ bufnr = 0 }` or `{ bufnr = api.nvim_get_current_buf() }`

Reply:
I guess then we also want `nil` range to imply the whole buffer (for each buffer). I'll work on that

Reply:
Actually, multiple buffers is trickier than I thought.

The [`InlayHint` struct](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#inlayHint) doesn't have documentURI or anything. So for multiple buffers to make sense, we'd have to add bufnr to the inlay hint object ourselves (or create a mapping). Is it worth it? What's the use case for multiple buffers?

Reply:
I think a simple usecase would be to have two windows open next to each other with different buffers.

Reply:
> Is it worth it? What's the use case for multiple buffers?

I don't have a concrete use-case in mind. It was mostly to go with the more flexible variant to avoid driving us into a corner with the choice.

Thinking about this some more, we may actually need to wrap the InlayHint structure into a custom table anyway to provide the client or client_id. Otherwise you can't reliably execute a command (as users shouldn't attempt to use _all_ connected clients to execute it, but the client which provided the inlayhint). Same would be the case for resolve (although that's still tbd how we deal with that in general)

Another option could be to go with `.get(bufnr, opts)` to make `bufnr` mandatory. No strong preference, but I'd tend to custom struct, and support all buffers, given the `client_id` requirement.

Reply:
Getting the hints might be helpful in case a plugin wants to trigger an `inlayHint/refresh` request, but that would require direct communication with the client (or the plugin being the LSP client itself).

Reply:
I ran into the `client_id` myself with the plugin demo. Sounds good. I've made the struct change.